pyscriptjs/src/components/pyrepl.ts

```diff
@@ -5,161 +5,201 @@ import { Compartment, StateCommand } from '@codemirror/state';
 import { keymap } from '@codemirror/view';
 import { defaultKeymap } from '@codemirror/commands';
 import { oneDarkTheme } from '@codemirror/theme-one-dark';
-import { getAttribute, addClasses, htmlDecode } from '../utils';
-import { BaseEvalElement } from './base';
+
+import { getAttribute,  ensureUniqueId, htmlDecode } from '../utils';
 import type { Runtime } from '../runtime';
+import { pyExec, pyDisplay } from '../pyexec';
 import { getLogger } from '../logger';
 
 const logger = getLogger('py-repl');
 
 export function make_PyRepl(runtime: Runtime) {
 
-    function createCmdHandler(el: PyRepl): StateCommand {
-        // Creates a codemirror cmd handler that calls the el.evaluate when an event
-        // triggers that specific cmd
-        return () => {
-            void el.evaluate(runtime);
-            return true;
-        };
-    }
-
-    let initialTheme: string;
-    function getEditorTheme(el: BaseEvalElement): string {
-        const theme = getAttribute(el, 'theme');
-        if( !initialTheme && theme){
-            initialTheme = theme;
-        }
-        return initialTheme;
-    }
-
-    class PyRepl extends BaseEvalElement {
+    /* High level structore of py-repl DOM, and the corresponding JS names.
+
+           this             <py-repl>
+           shadow               #shadow-root
+                                    <slot></slot>
+           boxDiv               <div class='py-repl-box'>
+           editorLabel              <label>...</label>
+           editorDiv                <div class="py-repl-editor"></div>
+           outDiv                   <div class="py-repl-output"></div>
+                                </div>
+                            </py-repl>
+    */
+    class PyRepl extends HTMLElement {
+        shadow: ShadowRoot;
+        outDiv: HTMLElement;
         editor: EditorView;
-        editorNode: HTMLElement;
 
         constructor() {
             super();
-
-            // add an extra div where we can attach the codemirror editor
-            this.editorNode = document.createElement('div');
-            addClasses(this.editorNode, ['editor-box']);
-            this.shadow.appendChild(this.wrapper);
         }
 
         connectedCallback() {
-            this.checkId();
-            this.code = htmlDecode(this.innerHTML);
+            ensureUniqueId(this);
+            this.shadow = this.attachShadow({ mode: 'open' });
+            const slot = document.createElement('slot');
+            this.shadow.appendChild(slot);
+
+            if (!this.hasAttribute('exec-id')) {
+                this.setAttribute('exec-id', '1');
+            }
+            if (!this.hasAttribute('root')) {
+                this.setAttribute('root', this.id);
+            }
+
+            const pySrc = htmlDecode(this.innerHTML).trim();
```

Comment:
Since you are touching this, we should also allow source to be defined form a `src` file and not only inline.

Reply:
Actually I already started to add it, but then I stopped because it raises too many open questions:
1. what happens if I cannot fetch the src? Where do I display the error?
2. what happens if I have `src` but also code inside the tag itself? I think it should be an error but we might want to discuss more.

Basically, the goal of this PR is not to fix every problem with py-repl but to clean up a bit of mess and make the code more readable/understandable. But I agree we should do more work in subsequent PRs (possibly not by me 😅)

(And semi-related, I'm also starting to think that we should change the current semantics of `<py-script src=...>`, and/or possibly change the name, because the current semantics is too fragile IMHO. But that's a topic for another dicussion).

Reply:
Ok, let's cover this in a follow-up PR.

> (And semi-related, I'm also starting to think that we should change the current semantics of <py-script src=...>, and/or possibly change the name, because the current semantics is too fragile IMHO. But that's a topic for another discussion).

Yeah, definitely a different discussion :) 👍